flink-core/src/main/java/org/apache/flink/api/common/typeutils/TypeSerializerConfigSnapshotSerializationUtil.java

```diff
@@ -0,0 +1,172 @@
+	 * written using {@link TypeSerializerConfigSnapshotSerializationUtil#writeSerializerConfigSnapshot(DataOutputView, TypeSerializerConfigSnapshot, TypeSerializer)}.
+	 *
+	 * @param in the data input view
+	 * @param userCodeClassLoader the user code class loader to use
+	 *
+	 * @return the read serializer configuration snapshots
+	 *
+	 * @throws IOException
+	 */
+	public static List<TypeSerializerConfigSnapshot<?>> readSerializerConfigSnapshots(
+			DataInputView in,
+			ClassLoader userCodeClassLoader) throws IOException {
+
+		int numFields = in.readInt();
+		final List<TypeSerializerConfigSnapshot<?>> serializerConfigSnapshots = new ArrayList<>(numFields);
+
+		TypeSerializerConfigSnapshotSerializationProxy proxy;
+		for (int i = 0; i < numFields; i++) {
+			proxy = new TypeSerializerConfigSnapshotSerializationProxy(userCodeClassLoader);
+			proxy.read(in);
+			serializerConfigSnapshots.add(proxy.getSerializerConfigSnapshot());
+		}
+
+		return serializerConfigSnapshots;
+	}
+
+	/**
+	 * Utility serialization proxy for a {@link TypeSerializerConfigSnapshot}.
+	 */
+	static final class TypeSerializerConfigSnapshotSerializationProxy<T> extends VersionedIOReadableWritable {
+
+		private static final int VERSION = 1;
+
+		private ClassLoader userCodeClassLoader;
+		private TypeSerializerConfigSnapshot<T> serializerConfigSnapshot;
+		private TypeSerializer<T> serializer;
+
+		TypeSerializerConfigSnapshotSerializationProxy(ClassLoader userCodeClassLoader) {
+			this.userCodeClassLoader = Preconditions.checkNotNull(userCodeClassLoader);
+		}
+
+		TypeSerializerConfigSnapshotSerializationProxy(
+			TypeSerializerConfigSnapshot<T> serializerConfigSnapshot,
+			TypeSerializer<T> serializer) {
+			this.serializerConfigSnapshot = Preconditions.checkNotNull(serializerConfigSnapshot);
+			this.serializer = Preconditions.checkNotNull(serializer);
+		}
+
+		@Override
+		public void write(DataOutputView out) throws IOException {
+			super.write(out);
+
+			// config snapshot class, so that we can re-instantiate the
+			// correct type of config snapshot instance when deserializing
+			out.writeUTF(serializerConfigSnapshot.getClass().getName());
+
+			// TODO this is a temporary workaround until all serializer config snapshot classes in Flink
+			// TODO have properly implemented the restore serializer factory methods
+			serializerConfigSnapshot.setSerializer(serializer);
+
+			// the actual configuration parameters
+			serializerConfigSnapshot.write(out);
+		}
+
+		@SuppressWarnings("unchecked")
+		@Override
+		public void read(DataInputView in) throws IOException {
+			super.read(in);
+
+			String serializerConfigClassname = in.readUTF();
+			Class<? extends TypeSerializerConfigSnapshot> serializerConfigSnapshotClass;
+			try {
+				serializerConfigSnapshotClass = (Class<? extends TypeSerializerConfigSnapshot>)
+					Class.forName(serializerConfigClassname, true, userCodeClassLoader);
```

Comment:
Minor: I found it often better to not initialize the class when reflectively loading it (pass false instead of true). Initialization will then happen on first access.

That means that errors during static initialization happen not in the deserialization stack, but upon use of the class later, which is often a bit easier to understand.